### ISSUE TITLE
mode() doesn't return Operator Pending mode?

### ISSUE THREAD

**Issue by peterwu** [vim/vim#12012]
### Steps to reproduce

1. put the following in .vimrc

```
set laststatus=2
set statusline=%{mode(1)}
```

2. launch vim
3. try to invoke the operator pending mode, such as yiw, daw, etc.
4. mode(1) doesn't return the operator pending mode as suggested by **:help mode()**  

### Expected behaviour

Any of the following should be returned for operator pending mode according to **:help mode()**

```
mode([expr])    Return a string that indicates the current mode.
                If [expr] is supplied and it evaluates to a non-zero Number or
                a non-empty String (non-zero-arg), then the full mode is
                returned, otherwise only the first letter is returned.
                Also see state().

                   n        Normal
                   no       Operator-pending
                   nov      Operator-pending (forced characterwise o_v)
                   noV      Operator-pending (forced linewise o_V)
                   noCTRL-V Operator-pending (forced blockwise o_CTRL-V);
                                CTRL-V is one character

```

### Version of Vim

9.0.1313

### Environment

OS: Windows
Terminal: N/A (using gVim)
$TERM: N/A
shell: N/A

OS: Fedora 37
Terminal: Xterm/ WSL on Windows Terminal
$TERM: xterm-256color
shell: bash 5.2.15

**Comment by numEricL** [comment]
Some added context, this is an issue only when setting the statusline. From `:help mode`


```
This is useful in the 'statusline' option

```

**Comment by brammool** [comment]
> ### Steps to reproduce
> 
> 1. put the following in .vimrc
> 
> ```
> set laststatus=2
> set statusline=%{mode(1)}
> ```
> 
> 2. launch vim
> 3. try to invoke the operator pending mode, such as yiw, daw, etc.
> 4. mode(1) doesn't return the operator pending mode as suggested by **:help mode()**  
> 
> ### Expected behaviour
> 
> Any of the following should be returned for operator pending mode according to **:help mode()**
> 
> ```
> mode([expr])    Return a string that indicates the current mode.
>                 If [expr] is supplied and it evaluates to a non-zero Number or
>                 a non-empty String (non-zero-arg), then the full mode is
>                 returned, otherwise only the first letter is returned.
>                 Also see state().
> 
>                    n        Normal
>                    no       Operator-pending
>                    nov      Operator-pending (forced characterwise o_v)
>                    noV      Operator-pending (forced linewise o_V)
>                    noCTRL-V Operator-pending (forced blockwise o_CTRL-V);
>                                 CTRL-V is one character
> 
> ```

The problem is not that mode() is returning a wrong value, the problem
is that the statusline is not redrawn.  A trick to have it redraw is
using this autocommand:

	:au CursorMoved * echowin 'mode:' mode(1)

Move the cursor over some text, the messages window will appear, on top
of the statusline.  Now start an operator, e.g. with "c" and wait a
couple of seconds.  The moment the messages window disappears the
statusline is redrawn and shows "no".

I'm not sure if it makes sense to redraw the statusline every time the
mode changes, just in case someone uses mode() in 'statusline'.  It adds
overhead, while only very few users would profit from it.

-- 
hundred-and-one symptoms of being an internet addict:
153. You find yourself staring at your "inbox" waiting for new e-mail
     to arrive.

 /// Bram Moolenaar -- ***@***.*** -- http://www.Moolenaar.net   \\\
///                                                                      \\\
\\\        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by peterwu** [comment]
> I'm not sure if it makes sense to redraw the statusline every time the
> mode changes, just in case someone uses mode() in 'statusline'.  It adds
> overhead, while only very few users would profit from it.

Agreed and I suggest `:help mode()` be updated to clarify as its current wording can be interpreted that, when setting 'statusline' option, a call to mode() will return a string indicating any of the modes listed and the string will be visible right away on the statusline just like any other mode, while, in fact, for operator pending modes a `redrawstatus` call must be made for the reasons you outlined.

```
mode([expr])    Return a string that indicates the current mode.
                If [expr] is supplied and it evaluates to a non-zero Number or
                a non-empty String (non-zero-arg), then the full mode is
                returned, otherwise only the first letter is returned.
                Also see state().

                   n        Normal
                   no       Operator-pending
                   nov      Operator-pending (forced characterwise o_v)
                   noV      Operator-pending (forced linewise o_V)
                   noCTRL-V Operator-pending (forced blockwise o_CTRL-V);
                                CTRL-V is one character
                   [...]

                This is useful in the 'statusline' option or when used
                with remote_expr() In most other places it always returns
                "c" or "n".
```
